Check in shim typeforwards and remove the dependency on the underlying targeting packs

ViktorHofer opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/65997

This also changes the folder structure of shims to make them fit into the existing libraries layout (projectname/ref, projectname/src). That makes sharing msbuild data and C# files easier and enables the use of default compile items.

The shim projects now also define their dependencies explicitly instead of just receiving all references.

ViktorHofer committed:
Make .NET Framework and .NET Standard facade changes trackable

Fixes https://github.com/dotnet/runtime/issues/65997

GenFacade is a tool that inspects a contract assembly's public API surface safe area, compares it with a passed in assembly closure and emits a C# source file with type forwards in it so that the satisfied API can be forwarded from the contract to the "implementation". GenFacades is used to construct the .NETFramework and the .NETStandard shim assemblies which live under src/libraries/shims/.

The generated source file isn't checked into the tree, it's placed into the project's intermediate output path. IMO it would help if we check these files into the tree so that any changes that impact shims - intentional or unintentional - are trackable. As an example, I just recently refactored how the shims are structured under src/libraries/shims and unintentionally removed a type forward from an assembly. I only noticed this regression by pure luck later.

The priority of this issue just rose because source build plans to remove their .NET Framework reference assembly packages during the .NET 8 release. That means that we can't automatically generate the type forwards during source-build. I will work on removing the dependency and check the type forwards in instead.

ericstj reviewed (commented):
I'm supportive of checking in the forwards, but I think we need to think about the workflow of this and expectations for how developers interact with these sources.  Consider adding a readme describing the interaction.

ViktorHofer committed:
Make shim projects declare the dependencies explicitly

Also remove the System.Xml special runtime project which isn't necessary
anymore as the compiler now allows type forwards to Obsolete types with
error=true.

ViktorHofer committed:
System.Core build fix

carlossanlop reviewed (commented):
I mostly have questions to learn about this.

Did the GenFacade tool generate both the cs files and csproj files under shims?

ViktorHofer commented on src/libraries/shims/Directory.Build.props:
```diff
@@ -19,44 +18,20 @@
-  </ItemGroup>
-
-  <ItemGroup>
-    <!-- System.Data.SqlClient is not live built anymore, reference it manually to avoid tfm nuget selection fallback errors
```
This reference was moved directly into shims/System.Data/Directory.Build.props

ViktorHofer commented on src/libraries/shims/System/Directory.Build.props:
```diff
@@ -1,6 +1,10 @@
-<Project Sdk="Microsoft.NET.Sdk">
+<Project>
```
That's just git that gets confused with the file changes...

ViktorHofer committed:
Remove unused file

ViktorHofer commented on src/libraries/shims/Directory.Build.props:
```diff
@@ -8,6 +8,8 @@
     <!-- By default make shims compile against reference assemblies. -->
     <CompileUsingReferenceAssemblies>true</CompileUsingReferenceAssemblies>
     <IsPartialFacadeAssembly>true</IsPartialFacadeAssembly>
+    <GenFacadesOutputSourcePath>$(MSBuildProjectDirectory)\$(MSBuildProjectName).g.cs</GenFacadesOutputSourcePath>
```
Sorry Eric, I should have changed the PR's state to WIP. You reviewed it a little bit too early. It should now be ready.

ViktorHofer committed:
Revert "Remove unused file"

This reverts commit 2f5d410a1ff2ed4705a7256c19cd65aa11107267.

ViktorHofer committed:
Add README and fix GenFacades zero version logic

carlossanlop commented on src/libraries/shims/Directory.Build.props:
```diff
@@ -4,59 +4,39 @@
-    <GenFacadesForceZeroVersionSeeds Condition="'$(MSBuildProjectName)' != 'netstandard'">true</GenFacadesForceZeroVersionSeeds>
+    <!-- Ensure the .NET Framework shims reference the lowest possible version of dependencies since
+         those do not all ship as part of the framework and we don't want to force apps to reference the
+         latest packages. netstandard.dll doesn't need to do this since it has no dangling dependencies. -->
```
Why do we not want the latest package to he referenced?

ViktorHofer replied:
This is already the same in main. Or are you asking about why we are doing this generally? IMO the comment captures that nicely. We make sure that shims don't depend on the very latest packages and instead work with any version. When you look at any of the shims in ilspy, i.e. mscorlib you will see that all dependency versions in its metadata are 0.0.0.0.

carlossanlop commented on src/libraries/shims/System.Xml.Serialization/src/System.Xml.Serialization.csproj:
```diff
@@ -0,0 +1,14 @@
+
+  <PropertyGroup>
+    <AssemblyVersion>4.0.0.0</AssemblyVersion>
+    <StrongNameKeyId>ECMA</StrongNameKeyId>
```
I see we use Microsoft, MicrosoftShared or ECMA values. What is the purpose of using difference values? How do you decide which one to use?

ViktorHofer replied:
These assemblies exist in the .NET Framework runtime on Windows machines and need to have the same identity. The strong name key is part of the assembly's identity, same as the version. We just pick whatever strong name key was chosen for the assembly in the .NET Framework runtime. Note that this also already existed in main, I didn't change any of these values, just moved them. Git diff unfortunately doesn't help here.

carlossanlop reviewed (approved):
Assuming the CI does not contain related failures, LGTM.
